### PR TITLE
Google Docs Integration

### PR DESCRIPTION
**Pull request by amitdavidson234**
## Status
Ready

## Related Issues
https://github.com/demisto/etc/issues/16140

## Description
Google Docs integration

### CODE REVIEW

File: Integrations/GoogleDocs/GoogleDocs.py [hunk 1]
```diff
@@ -0,0 +1,395 @@
+import demistomock as demisto
```
Comment (hirshbergshachar) [inline Integrations/GoogleDocs/GoogleDocs.py]: Remove those 3 lines

File: Integrations/GoogleDocs/GoogleDocs.py [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+from googleapiclient import discovery
+
+import typing
+from collections import defaultdict
+# Disable insecure warnings
+requests.packages.urllib3.disable_warnings()
+
+''' GLOBALS/PARAMS '''
+
+SCOPES = ['https://www.googleapis.com/auth/documents']
```
Comment (hirshbergshachar) [inline Integrations/GoogleDocs/GoogleDocs.py]: What does this mean? Can you add a comment to explain?

File: Integrations/GoogleDocs/GoogleDocs.py [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+
+ACTION_TO_FUNCTION = {
```
Comment (hirshbergshachar) [inline Integrations/GoogleDocs/GoogleDocs.py]: Nice

File: Integrations/GoogleDocs/GoogleDocs.py [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+    actions = actions.split(';')
+    for action in actions:
+        action_type, params = action.split('{')
+        params = params[:-1]
+        params = params.split(',')
+        parsed_actions[action_type] = params
+    return parsed_actions
+
+
+def log_error(f):
```
Comment (hirshbergshachar) [inline Integrations/GoogleDocs/GoogleDocs.py]: @SgtJahnoon very nice.
@Itay4 do we have such decorator already? If not perhaps we can add it to common server python

File: Integrations/GoogleDocs/GoogleDocs.py [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+        except Exception as e:
+            return_error(str(e))
+
+    return wrapped
+
+
+@log_error
+def get_http_client_with_proxy(disable_ssl):
+    proxies = handle_proxy()
+    if not proxies or not proxies['https']:
```
Comment (hirshbergshachar) [inline Integrations/GoogleDocs/GoogleDocs.py]: Please double check the logic

Comment (hirshbergshachar) [reply Integrations/GoogleDocs/GoogleDocs.py]: I don't see when `proxies` can equal `False`

File: Integrations/GoogleDocs/GoogleDocs.py [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+        raise Exception('https proxy value is empty. Check Demisto server configuration')
+    https_proxy = proxies['https']
```
Comment (hirshbergshachar) [inline Integrations/GoogleDocs/GoogleDocs.py]: Use `get`?

File: Integrations/GoogleDocs/GoogleDocs.py [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+
+    if proxy or disable_ssl:
+        http_client = credentials.authorize(get_http_client_with_proxy(disable_ssl))
+        return discovery.build('docs', 'v1', http=http_client)
+    return discovery.build('docs', 'v1', credentials=credentials)
+
+
+''' COMMANDS + REQUESTS FUNCTIONS '''
+
+
```
Comment (hirshbergshachar) [inline Integrations/GoogleDocs/GoogleDocs.py]: Good helper functions

File: Integrations/GoogleDocs/GoogleDocs.py [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+    elif target_revision_id:
+        write_control['writeControl']["targetRevisionId"] = target_revision_id
+
+    payload = {**payload, **write_control}
+
+    for action_type, params in actions.items():
+        request = globals()[ACTION_TO_FUNCTION[action_type]](action_type, *params)
+        payload["requests"].append(request)
+
+    service.documents().batchUpdate(documentId=document_id, body=payload).execute()
```
Comment (hirshbergshachar) [inline Integrations/GoogleDocs/GoogleDocs.py]: Can we get an error here?

File: Integrations/GoogleDocs/GoogleDocs.py [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+
+def get_document_command(service):
+    args = demisto.args()
+    document_id = args.get('document_id')
+    document = get_document(service, document_id)
+    return document
+
+
+def get_document(service, document_id):
+    document = service.documents().get(documentId=document_id).execute()
```
Comment (hirshbergshachar) [inline Integrations/GoogleDocs/GoogleDocs.py]: Can we have an error here?

File: Integrations/GoogleDocs/GoogleDocs.py [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+    return document
+
+
+def main():
+    LOG('Command being called is %s' % (demisto.command()))
```
Comment (hirshbergshachar) [inline Integrations/GoogleDocs/GoogleDocs.py]: Should be on debug log

File: Integrations/GoogleDocs/GoogleDocs.py [hunk 11]
```diff
@@ -0,0 +1,395 @@
...
+    disable_ssl = demisto.params().get('insecure', False)
+    service_account_credentials = json.loads(demisto.params().get('service_account_credentials'))
+    if demisto.command() == 'test-module':
+        try:
+            get_client(service_account_credentials, SCOPES, proxy, disable_ssl)
+            demisto.results('ok')
+        except Exception as e:
+            return_error("Failed to execute test. Error: {}".format(str(e)), e)
+
+    try:
```
Comment (hirshbergshachar) [inline Integrations/GoogleDocs/GoogleDocs.py]: Good

File: Integrations/GoogleDocs/GoogleDocs.py [hunk 12]
```diff
@@ -0,0 +1,395 @@
...
+        elif demisto.command() == 'create_document':
+            res = create_document_command(service)
+        elif demisto.command() == 'get_document':
+            res = get_document_command(service)
+        else:
+            return_error("Command {} does not exist".format(demisto.command()))
+            return
+
+        ec = {
+            'GoogleDocs(val.DocumentId && val.DocumentId == DocumentId.Query)': {
```
Comment (hirshbergshachar) [inline Integrations/GoogleDocs/GoogleDocs.py]: Why do we compare the `DocumentId` to the `Query`?

File: Integrations/GoogleDocs/GoogleDocs.py [hunk 13]
```diff
@@ -0,0 +1,395 @@
...
+                'DocumentId': res['documentId'],
+                'Title': res['title']
+            }
+        }
+
+        demisto.results({
+            'Type': entryTypes['note'],
+            'ContentsFormat': formats['markdown'],
+            'Contents': ec,
+            'HumanReadable': json.dumps(ec),
```
Comment (hirshbergshachar) [inline Integrations/GoogleDocs/GoogleDocs.py]: We'll probably need to make a better human readable, let's revisit this point after we'll sit on the integration together

**Comment by hirshbergshachar** [comment]
@kirbles19 please make sure the integration text is reviewed

**Comment by SgtJahnoon** [comment]
@kirbles19 These are screenshots of the commands. 
Screen Shot 2019-09-26 at 16 31 07
Screen Shot 2019-09-26 at 16 31 14
Screen Shot 2019-09-26 at 16 31 21

**Comment by kirbles19** [comment]
@amitdavidson234 Done.

**Pull request merged**